Undefined symbols from pthread in empty staticlib for wasip1

<!--
Thank you for filing a regression report! 🐛 A regression is something that changed between versions of Rust but was not supposed to.

Please provide a short summary of the regression, along with any information you feel is relevant to replicate it.
-->

### Code

I tried this code:

```rust
```

I expected to see this happen: compiling an empty lib with crate type `staticlib` to target `wasm32-wasi` creates a relocatable Wasm `.a` without undefined symbols for pthreads.

Instead, this happened: the `.a` contains undefined symbols for pthreads (the `.wasm` has no undefined symbols)

[repro.sh](https://github.com/user-attachments/files/27210903/repro.sh)

```
rustc 1.93.1 (01f6ddf75 2026-02-11)
=== undefined pthread symbols ===

rustc 1.94.1 (e408947bf 2026-03-25)
=== undefined pthread symbols ===
--- std-c91e563353c7e174.std.728e8bab36f00-cgu.0.rcgu.o ---
   - 454: F <env.pthread_join> func=91 [ undefined binding=global vis=default ]
   - 1278: F <env.pthread_detach> func=142 [ undefined binding=global vis=default ]
 
rustc 1.95.0 (59807616e 2026-04-14)
=== undefined pthread symbols ===
--- std-d64733c6df5094a8.std.a88812da5ee6ad9e-cgu.0.rcgu.o ---
   - 281: F <env.pthread_attr_init> func=52 [ undefined binding=global vis=default ]
   - 282: F <env.pthread_attr_setstacksize> func=53 [ undefined binding=global vis=default ]
   - 285: F <env.pthread_attr_destroy> func=55 [ undefined binding=global vis=default ]
   - 292: F <env.pthread_create> func=56 [ undefined binding=global vis=default ]
   - 294: F <env.pthread_join> func=57 [ undefined binding=global vis=default ]
   - 819: F <env.pthread_detach> func=114 [ undefined binding=global vis=default ]
   - 
rustc 1.97.0-nightly (37d85e592 2026-04-28)
=== undefined pthread symbols ===
--- std-e31abec1fa1087a7.std.22fe8ebee811282d-cgu.0.rcgu.o ---
   - 281: F <env.pthread_attr_init> func=52 [ undefined binding=global vis=default ]
   - 282: F <env.pthread_attr_setstacksize> func=53 [ undefined binding=global vis=default ]
   - 285: F <env.pthread_attr_destroy> func=55 [ undefined binding=global vis=default ]
   - 292: F <env.pthread_create> func=56 [ undefined binding=global vis=default ]
   - 294: F <env.pthread_join> func=57 [ undefined binding=global vis=default ]
   - 826: F <env.pthread_detach> func=114 [ undefined binding=global vis=default ]
```

Triggered by https://github.com/rust-lang/rust/pull/153634?

### Version it worked on

<!--
Provide the most recent version this worked on, for example:

It most recently worked on: Rust 1.47
-->

It most recently worked on: 1.93

### Version with regression

<!--
Provide the version you are using that has the regression.
-->

`rustup run 1.94 rustc --version --verbose`:
```
rustc 1.94.1 (e408947bf 2026-03-25)
binary: rustc
commit-hash: e408947bfd200af42db322daf0fadfe7e26d3bd1
commit-date: 2026-03-25
host: aarch64-apple-darwin
release: 1.94.1
LLVM version: 21.1.8
```

`rustup run nightly rustc --version --verbose`:
```
rustc 1.97.0-nightly (37d85e592 2026-04-28)
binary: rustc
commit-hash: 37d85e592f9ae5f20f7d9a9f99785246fa7298da
commit-date: 2026-04-28
host: aarch64-apple-darwin
release: 1.97.0-nightly
LLVM version: 22.1.4
```

<!--
If you know when this regression occurred, please add a line like below, replacing `{channel}` with one of stable, beta, or nightly.

@rustbot modify labels: +regression-from-stable-to-{channel} -regression-untriaged
-->

@benozol can you explain at a higher level why you're expecting this or what you're hoping to achieve? This is expected behavior because Rust `*.a` staticlib files still depend on libc/standard APIs. For example if you compile for a native platform and produce a staticlib it's going to have undefined symbols for functions/data found in `libc.so` on your native system. (or something else on Windows).

The undefined `pthread_*` symbols are functions that the standard library is using from `libc.a`, provided in wasi-sdk as a standard wasi toolchain. Rust also ships with its own `libc.a`, but that's not included in a Rust staticlib since it's a "self contained sysroot" onl used during linking.

--- a/src/doc/rustc/src/platform-support/wasm32-wasip1.md
+++ b/src/doc/rustc/src/platform-support/wasm32-wasip1.md
@@ -70,17 +70,16 @@ be used instead.
 
 [`wasi-libc`]: https://github.com/WebAssembly/wasi-libc
 
-## Building the target
+## Building the target in rustc
 
-To build this target first acquire a copy of
-[`wasi-sdk`](https://github.com/WebAssembly/wasi-sdk/). At this time version 22
-is the minimum needed.
+To build this target first acquire a copy of [`wasi-sdk`]. At this time version
+33 is the minimum needed.
 
 Next configure the `WASI_SDK_PATH` environment variable to point to where this
 is installed. For example:
 
 ```text
-export WASI_SDK_PATH=/path/to/wasi-sdk-22.0
+export WASI_SDK_PATH=/path/to/wasi-sdk-33.0
 ```
 
 Next be sure to enable LLD when building Rust from source as LLVM's `wasm-ld`
@@ -102,6 +101,16 @@ Rust programs can be built for that target:
 ```text
 rustc --target wasm32-wasip1 your-code.rs
 ```
+
+The `wasm32-wasip1` toolchain comes with a self-contained sysroot meaning that
+no external compiler is required when building for this target. Users which
+build a `staticlib`, however, or use an external `-Clinker`, are recommended to
+use [`wasi-sdk`]. The minimum version required of [`wasi-sdk`] will change over
+time as it's updated in Rust and Rust relies on newer features that [`wasi-sdk`]
+has. See the documentation above about building the target in rustc for the
+minimum known working version.
+
+[`wasi-sdk`]: https://github.com/WebAssembly/wasi-sdk
 
 ## Cross-compilation
 

--- a/src/doc/rustc/src/platform-support/wasm32-wasip2.md
+++ b/src/doc/rustc/src/platform-support/wasm32-wasip2.md
@@ -27,23 +27,20 @@ are required to support components since this target outputs a component as
 opposed to a core wasm module. As of the time of this writing Wasmtime 17 and
 above is able to run this target natively with no extra flags.
 
-## Building the target
+## Building the target in rustc
 
-To build this target first acquire a copy of
-[`wasi-sdk`](https://github.com/WebAssembly/wasi-sdk/). At this time version 22
-is the minimum needed.
+See the documentation for the [building the `wasm32-wasip1` target in
+rustc](./wasm32-wasip1.md#building-the-target-in-rustc) for more information. The tl;dr;
+is that [`wasi-sdk`] is required, and the `wasm32-wasip1` target documents the
+minimum version required.
 
-Next configure the `WASI_SDK_PATH` environment variable to point to where this
-is installed. For example:
+[`wasi-sdk`]: https://github.com/WebAssembly/wasi-sdk
 
-```text
-export WASI_SDK_PATH=/path/to/wasi-sdk-22.0
-```
+## Building Rust programs
 
-Next be sure to enable LLD when building Rust from source as LLVM's `wasm-ld`
-driver for LLD is required when linking WebAssembly code together. Rust's build
-system will automatically pick up any necessary binaries and programs from
-`WASI_SDK_PATH`.
+For more information see the documentation [`wasm32-wasip1`
+target](./wasm32-wasip1.md#building-rust-programs). Replace `wasm32-wasip1`
+target strings with `wasm32-wasip2`, however.
 
 ## Testing
 

